src/main/java/recipeapplication/database/RecipeFirebaseCrud.java

```diff
@@ -61,6 +65,36 @@ public Recipe delete(String document) {
         return deletedRecipe;
     }
 
+    public Recipe[] queryCategory(String category) {
+        CollectionReference recipes = firestore.collection(collection);
+        Query query = recipes.whereEqualTo("category", category.toUpperCase());
+
+        return getRecipeArrayFromQuery(query).toArray(new Recipe[0]);
+
+    }
+
+    public Recipe[] queryIngredients(String[] ingredients) {
+        CollectionReference recipes = firestore.collection(collection);
+        Query query = recipes.whereArrayContainsAny("ingredients", Arrays.asList(ingredients));
+
+        ArrayList<Recipe> recipeArrayBeforeFilter = getRecipeArrayFromQuery(query);
+        ArrayList<Recipe> recipeArrayAfterFilter = new ArrayList<>();
+        for (Recipe recipe : recipeArrayBeforeFilter) {
+            if(StringUtils.isFullyContainedInArray(recipe.getIngredients(), Arrays.asList(ingredients))) {
+                recipeArrayAfterFilter.add(recipe);
+            }
```

Comment:
This can be made much easier with streams and filters
`return getRecipeArrayFromQuery(query).stream().filter(recipe -> listToUpperCase(recipe.getIngredients()).containsAll(listToUpperCase(Arrays.asList(ingredients))).toArray(new Recipe[0])`

Reply:
Altered to 

return getRecipeArrayFromQuery(query).stream()
                .filter(recipe -> StringUtils.listToUpperCase(Arrays.asList(ingredients))
                        .containsAll(StringUtils.listToUpperCase(recipe.getIngredients())))
                .collect(Collectors.toList()).toArray(new Recipe[0]);